[Intel-Mkl] Redundant transpose removal: transpose(to "NHWC")+ conv2d(NHWC) + transpose(back to "NCHW") -> conv2d.

wenxizhu opened the pull request:
This is a pattern found in Keras-based models. When using "NCHW" format dataset, Keras will insert 2 extra "Transpose" ops before/after every CNN ops, because only "NHWC" format is supported in CPU mode. 

These "Transpose" overhead are significant in typical CNN models, we observed it contributes ~35% of total inference time in 3D-UNet. This PR is aimed to remove these overhead completely, by:
1. detect the "transpose + conv2d + transpose" pattern;
2. delete those 2 "transpose" ops, and make "conv2d"'s format "NCHW";
3. replace "conv2d" with "_MklConv2D", which support both "NCHW" and "NHWC" format.

wenxizhu committed:
Redundant transpose removal: transpose + conv2d + transpose -> conv2d.

martinwicke commented:
@wt-huang is the CLA ok?

martinwicke commented:
@wenxizhu have you signed the CLA? Github's been malfunctioning lately so I think our CLA checked ignored this PR.

wt-huang commented:
@martinwicke Thanks for the reminder!

@wenxizhu Please make sure CLA is signed so that we can proceed with this PR.

wenxizhu commented:
@martinwicke @wt-huang I'm pretty sure I have my CLA signed. What shall I do now?

wenxizhu commented:
@wt-huang seems it doesn't pass your "clang-format" test? I'm wondering how to run it on m local device. "//tensorflow/tools/ci_build:gen_ci_clang_format_out" is not a valid bazel so I could not run it directly. Any suggestions?

wenxizhu commented:
@wt-huang "no such package 'tensorflow/tools/ci_build': BUILD file not found on package path"

penpornk commented:
@wenxizhu You can use [clang-format](https://clang.llvm.org/docs/ClangFormat.html) given with the [LLVM toolchain](http://releases.llvm.org/download.html) to format the source files to pass the `clang-format` test.

wenxizhu committed:
Fix clang-format issues.

wenxizhu committed:
Fxi clang-format issues.

wenxizhu commented:
@penpornk I've applied "clang-format-3.8" to fix the problem. Can we move on now?

penpornk commented:
@wenxizhu Yes, I've been meaning to review this today but didn't get to. Sorry for the delay! I'll do it tomorrow.

penpornk commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -2391,6 +2646,150 @@ MklLayoutRewritePass::CheckForNodeRewrite(const Node* n) const {
+            if (check_next_node(candidate_node) == false) {
+              current_node = nullptr;
+            } else {
+              current_node = candidate_node;
```
Greedy matching like this could miss a pattern. For example, we are looking for a pattern X->Y->Z. Found a node X which has two output edges to two nodes Y1 and Y2. Y1 doesn't connect to Z. Y2 connects to Z. This loop would add Y1 into the array `nodes`, ignore Y2, and will eventually miss a match X->Y2->Z.

This issue won't happen with TransposeMklOpTranspose because the fusion rules only fuse when all the nodes only have one output data edge. I'm asking for a future/more generic case.

penpornk commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -2516,6 +2915,29 @@ bool MklLayoutRewritePass::RunPass(std::unique_ptr<Graph>* g) {
 
+  order.clear();
+  GetReversePostOrder(**g, &order);  // This will give us topological sort.
+  for (Node* n : order) {
```
This pass happening after node merging means it could miss an opportunity for a long fusion. For example, for a graph A->B->C->D->E and merge rules {A->B to AB, D->E to DE} and fusion rule {A->B->C->D->E to ABCDE}, the result will be AB->C->DE instead of ABCDE.

If you plan to do a long fusion chain only through merge rules, i.e., {A->B to AB, AB->C to ABC, ...} and only use fusion rules that doesn't contain any patterns already occurred in merge rules, please add a comment clarifying this.

penpornk commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -2391,6 +2646,150 @@ MklLayoutRewritePass::CheckForNodeRewrite(const Node* n) const {
+  std::vector<Node*> nodes;
+  const FusionInfo* fi_ptr = nullptr;
+
+  for (auto fi = finfo_.begin(); fi != finfo_.end(); ++fi) {
```
`fi` that shows up first will get applied first. This could result in a suboptimal fusion. For example, for a graph A->B->C->D and fusion rules {A->B->C to ABC, A->B->C->D to ABCD}, since the first rule gets applied first, the final graph will be ABC->D.

wenxizhu replied:
@penpornk yes, and it is also true for "merge". Optimal fusion can be achieved by carefully arrange the order of `fi`, put "important" fusion patterns at front of the list.

wenxizhu replied:
A comment added in the "finfo_.push_back(..." stage to clarify the behavior of this fusion pattern match process.

penpornk commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -2391,6 +2646,150 @@ MklLayoutRewritePass::CheckForNodeRewrite(const Node* n) const {
+  FillInputs(transpose_to_nhwc, &transpose_nchw_control_edges,
+             &transpose_nchw_in);
+
+  // We will use the node name of Conv2d as the name of new node
```
This line seems redundant with the two lines below. Remove?

wenxizhu replied:
@penpornk it should be `transpose_to_nchw`. Fixed the typo.

wenxizhu replied:
Ok, that's a happy coincidence I think, help me find a potential bug here... And the comments has been fixed now.

wenxizhu replied:
Correct, thank you:)

penpornk commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -2391,6 +2646,150 @@ MklLayoutRewritePass::CheckForNodeRewrite(const Node* n) const {
+  // We will use the node name of Conv2d as the name of new node
+  // Build new node. We use same name as original node, but change the op
+  // name.
+  NodeBuilder nb(mklop->name(), mklop->type_string());
```
How is the op name changed? We are still using `mklop->type_string()`.

wenxizhu replied:
@penpornk we won't change the op name, for example, "Conv2D" will still be "Conv2D", the only change is its `data_format`, from NHWC to NCHW.

wenxizhu replied:
Yes, fixed.

penpornk reviewed (changes requested):
Thank you for your PR and sorry again for the delay! I have a few questions about how the fusion rules are matched. It could miss some matches in a more general setting.

penpornk committed:
Replace "=>" with "into" for readability.

Co-Authored-By: wenxizhu <wenxi.zhu@intel.com>

penpornk committed:
Fix a grammatically incorrect in comments.

Co-Authored-By: wenxizhu <wenxi.zhu@intel.com>

wenxizhu committed:
Merge the line with its previous line.

wenxizhu commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -508,6 +511,36 @@ class MklLayoutRewritePass : public GraphOptimizationPass {
+    // NCHW) " => "Conv2D (NCHW). Such patterns occur frequently in Keras.
+    // Note: we use the term "merge" is to combine (exactly) 2 nodes into one,
+    // while "fusion" is
+    // for 3+ nodes situation.
```
@penpornk  fixed.

wenxizhu committed:
2+ should be 3+.

wenxizhu committed:
Capitalized the beginning of the sentences.

wenxizhu commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -560,6 +590,42 @@ class MklLayoutRewritePass : public GraphOptimizationPass {
     std::function<Node*(const Node*)> get_node_to_be_merged;
   } MergeInfo;
 
+  // structure to specify information used in node fusion of 2+ operators
```
@penpornk yes, fixed.

wenxizhu committed:
Remove empty comment lines.

wenxizhu commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -815,6 +885,121 @@ class MklLayoutRewritePass : public GraphOptimizationPass {
+      string data_format);
+
+  static bool CheckForTranspose(const Node *node, std::vector<int> perm) {
+    //
```
@penpornk These empty comment lines have been removed now.

penpornk committed:
Fix a grammatically incorrect comment line.

Co-Authored-By: wenxizhu <wenxi.zhu@intel.com>

penpornk committed:
Use "CHECK_EQ" instead of "CHECK".

Co-Authored-By: wenxizhu <wenxi.zhu@intel.com>

wenxizhu committed:
Explain the high level idea of comparing 'perm' and 'perm_node'.

wenxizhu commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -815,6 +885,121 @@ class MklLayoutRewritePass : public GraphOptimizationPass {
+    if (node->num_outputs() > 1 || node->out_edges().size() > 1)
+      return false;
+
+    // Check "perm" attribute, make sure it's what we want.
```
@penpornk done.

wenxizhu commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -815,6 +885,121 @@ class MklLayoutRewritePass : public GraphOptimizationPass {
+          //
+          // Here we directly access to the "tensor_context", rather than
+          // "int_val". This is because we find "int_val" is
+          // not set properly under some circumstances.
```
@penpornk according to my observation, in keras-based models, all "transpose" ops' int_val/int64_val/float_val... are not set entirely (all zero), so I have to parse the raw tensor data from "tensor_context" here myself. This is what I called "some circumstance" above. I don't think it's a desired behavior, and I will fire a bug to discuss about it.

penpornk committed:
Fix a typo.

Co-Authored-By: wenxizhu <wenxi.zhu@intel.com>

wenxizhu committed:
Remove some empty lines.

wenxizhu commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -815,6 +885,121 @@ class MklLayoutRewritePass : public GraphOptimizationPass {
+            return perm_value == long_perm;
+
+          }
+
```
@penpornk fixed.

wenxizhu commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -815,6 +885,121 @@ class MklLayoutRewritePass : public GraphOptimizationPass {
+        }
+      }
+    }
+
```
@penpornk fixed.

wenxizhu committed:
Check if node is nullptr.

wenxizhu commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -815,6 +885,121 @@ class MklLayoutRewritePass : public GraphOptimizationPass {
+  }
+
+  static bool CheckForMklOp(const Node *node, string name = "") {
+    if (!name.empty() && node->type_string() != name) {
```
@penpornk done.

wenxizhu committed:
Add description for FuseTransposeMklOpTranspose().

wenxizhu commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -815,6 +885,121 @@ class MklLayoutRewritePass : public GraphOptimizationPass {
+  Status FuseNode(std::unique_ptr<Graph> *g, std::vector<Node *> &nodes,
+                  const MklLayoutRewritePass::FusionInfo fi);
+
+  static Status FuseTransposeMklOpTranspose(
```
> Please add a function description.

Done.

wenxizhu committed:
Remove empty comment lines.

wenxizhu commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -1599,18 +1801,72 @@ void MklLayoutRewritePass::CopyAttrsConv(const Node* orig_node,
+    std::vector<int32> new_strides;
+    std::vector<int32> new_dilations;
+    if (strides.size() == 5) {
+      //
```
> Please remove blank comment lines in this function.

Removed.

wenxizhu commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -1599,18 +1801,72 @@ void MklLayoutRewritePass::CopyAttrsConv(const Node* orig_node,
+      // "strides" and "dilations" also need to be changed according to "data_format",
+      // in this case, is "NDHWC" to "NCDHW".
+      //
+
```
@penpornk removed.

wenxizhu committed:
Move nb->Attr() out of if statement.

wenxizhu commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -1599,18 +1801,72 @@ void MklLayoutRewritePass::CopyAttrsConv(const Node* orig_node,
+                      strides[NDHWC::dim::H],
+                      strides[NDHWC::dim::W]
+                    };
+      nb->Attr("strides", new_strides);
```
@penpornk done.

wenxizhu commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -1599,18 +1801,72 @@ void MklLayoutRewritePass::CopyAttrsConv(const Node* orig_node,
+
+    TF_CHECK_OK(GetNodeAttr(orig_node->def(), "data_format", &data_format));
+    nb->Attr("data_format", data_format);
+  } else {
```
@penpornk I'm afraid not. `nb->Attr("data_format", ...)` is only allowed to be called exactly once for a single nodebuilder. If we set `data_format` here, the next time we call `nb->Attr("data_format", ...)` will throw an error.

wenxizhu commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -2231,7 +2500,8 @@ Status MklLayoutRewritePass::RewriteNode(std::unique_ptr<Graph>* g,
 
-  ri->copy_attrs(const_cast<const Node*>(orig_node), &nb);
+  const bool kPartialCopyAttrs = false;
+  ri->copy_attrs(const_cast<const Node*>(orig_node), &nb, kPartialCopyAttrs);
```
@penpornk Yes, it is. With this `kPartialCopyAttrs` set to `true`, only those attributes NOT related to `data_format` will be copied. `data_format`, `strides` and `dilations` will be set in "FuseNode()", so no need to copy them. That's why I call it `kPartialCopyAttrs`.

penpornk committed:
Fix a variable name mis-spelling.

Co-Authored-By: wenxizhu <wenxi.zhu@intel.com>

wenxizhu commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -2391,6 +2661,151 @@ MklLayoutRewritePass::CheckForNodeRewrite(const Node* n) const {
+  gtl::InlinedVector<Node *, 4> transpose_nchw_control_edges;
+  gtl::InlinedVector<std::pair<Node *, int>, 4> transpose_nchw_in(
+      transpose_nchw_num_inputs);
+  FillInputs(transpose_to_nhwc, &transpose_nchw_control_edges,
```
@penpornk Yes, fixed.

wenxizhu commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -2391,6 +2661,151 @@ MklLayoutRewritePass::CheckForNodeRewrite(const Node* n) const {
+
+  // Fill outputs.
+  for (const Edge *e : transpose_to_nchw->out_edges()) {
+    if (!e->IsControlEdge()) {
```
@penpornk I did a check in "CheckForTranspose()" to make sure transpose ops have no control edges, so this won't be a problem.

wenxizhu committed:
Fix a typo.

wenxizhu commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -2391,6 +2661,151 @@ MklLayoutRewritePass::CheckForNodeRewrite(const Node* n) const {
+  }
+
+  // Copy device assigned to old node to new node.
+  new_node->set_assigned_device_name(mklop->assigned_device_name());
```
@penpornk that's a typo. Fixed.

wenxizhu replied:
@penpornk why `nb.Device(mklop->def().device())` is redundant? It didn't set device name of `mklop`.

wenxizhu replied:
@penpornk `requested_device` and `assigned_device_name_index` copied.

wenxizhu commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -2391,6 +2661,151 @@ MklLayoutRewritePass::CheckForNodeRewrite(const Node* n) const {
+  // Fill outputs.
+  for (const Edge *e : transpose_to_nchw->out_edges()) {
+    if (!e->IsControlEdge()) {
+      const int kConv2DWithBiasOutputSlot = 0;
```
> Why is this `WithBias`?

That's a type, fixed.

wenxizhu committed:
Copy requested_device and assigned_device_name_index.

wenxizhu committed:
Remove an assert line.

wenxizhu commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -2391,6 +2661,151 @@ MklLayoutRewritePass::CheckForNodeRewrite(const Node* n) const {
+  const FusionInfo *fi_ptr = nullptr;
+
+  for (auto fi = finfo_.begin(); fi != finfo_.end(); ++fi) {
+    assert(fi->ops.size() == fi->node_checkers.size());
```
@penpornk `ops` used to be a member of `FusionInfo` however I removed it in one revision. Anyway it shouldn't be here, you're right, I will remove this line.

wenxizhu committed:
Fix a typo.

wenxizhu committed:
Replace the greedy search algorithm in "CheckForNodeFusion()" with a stack-based one, to avoid missing some patterns.

penpornk commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -2729,65 +2719,70 @@ Status MklLayoutRewritePass::FuseNode(
       } else {
-        found_pattern = true;
+        // current node doesn't match, just break and stack will help us roll back.
+        break;
```
This will `break` the `while (!work_stack.empty())` loop. I don't think it will be able to find patterns like `A->B->C` (there is an example with 2 nodes as well but we are talking about fusion so I make it three) in a graph with edges `A->C`, `A->B`, and `B->C`. `C` is the destination of the first outgoing edge from `A`. It will be top of stack, won't pass `*node_checker`, will `break` the `while` loop, and go to a new `fi`.

wenxizhu replied:
You're right, this "else" branch was not handled properly. A correct behavior should be:
1. pop the stack;
2. mark current node as "visited";
3. decrease "node_checker".

Above change has been applied.

wenxizhu replied:
And there is a subtle bug in the if statement when `node_checker` returns true case. The nested level of stack pop was also wrong, it should be after the for loop, rather than in the loop. Fixed now.

penpornk reviewed (changes requested):
Thank you for the edits! I still don't think `CheckForNodeFusion` is working as intended.

wenxizhu committed:
Add a comment to the constraint of "inference-only" to note it will eventually be removed, if we enabled this fusion for training in the future.

wenxizhu commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -898,43 +899,38 @@ class MklLayoutRewritePass : public GraphOptimizationPass {
     // Check if has out control edge. If true, this is a training graph.
     // Currently we focus on inference and do no fusion in training.
-    //
     for (const Edge* e : node->out_edges()) {
```
Comments added.

wenxizhu committed:
Move a cheaper early return to the top of the function.

wenxizhu committed:
Set strides outside of if-else.

wenxizhu commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -1833,23 +1823,19 @@ void MklLayoutRewritePass::CopyAttrsConv(const Node* orig_node, NodeBuilder* nb,
 
       new_strides = {strides[NHWC::dim::N], strides[NHWC::dim::C],
                      strides[NHWC::dim::H], strides[NHWC::dim::W]};
       nb->Attr("strides", new_strides);
```
Done.

wenxizhu committed:
Fix a comment line.

wenxizhu committed:
Add a comment to clarify that patterns in finfo_ shows up first will get applied first.

wenxizhu commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -2729,65 +2719,70 @@ Status MklLayoutRewritePass::FuseNode(
-        break;
-      }
+    std::stack<Node *, std::vector<Node *>> work_stack;
+    std::set<Node *> visited_nodes;
```
True. Fixed now.

wenxizhu committed:
Use std::unordered_set instead of std::set, for better performance.

wenxizhu committed:
Fix 2 bugs in CheckNodeForFusion():
1. The else case for node_checker fails are not handled properly. Should pop the stack, rather then break the while loop entirely.
2. The nested level of stack op when node_check succeed is wrong.

wenxizhu commented:
@penpornk thanks for your comments! Changes have been applied, especially for the `CheckForNodeFusion`. And again, please let me know if I missed any of your comments.

penpornk commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -2391,6 +2638,161 @@ MklLayoutRewritePass::CheckForNodeRewrite(const Node* n) const {
+    //
+
+    std::stack<Node *, std::vector<Node *>> work_stack;
+    std::unordered_set<Node *> visited_nodes;
```
Using a global `visited_nodes` set is not enough. We can use the same graph as last time for example: fusion pattern `A->B->C` with graph `A->C`, `A->B`, and `B->C`. `C` is visited first and marked as a `visited` node, so by the time we reach `B`, we won't go to `C` anymore. (This case seems relevant enough. Please also add a test case for this.)

I also don't like that we have to iterate pass the already processed `current_node->out_edges()` every time (even though all we do with them is a quick check). Instead of a `visited_nodes` set, you can just save the current out edge for each checker index, `0 <= index < node_checkers.size()`. The next time you revisit this checker index level you can just go to the next edge. You can also get rid of the `work_stack` and store the results directly in `nodes` to avoid having to copy it at the end. For example:
```c++
std::vector</*EdgeSet iterator type*/> current_neighbor(fi->node_checkers.size());
// To get the candidate node inside the loop
candidate_node = current_neighbor[index];
// To have the next candidate ready
++current_neighbor[index];
// To store the candidate node.
nodes[index] = candidate_node;
// To check if you're out of neighbor
if (current_neighbor[index] == nodes[index-1]->out_edges().end())
```

wenxizhu replied:
You're right, that's a good counter-example. My algorithm has implications that subgraph is always a tree (whose root is "A"), that may not be true in real-world models. 

I modified the code, applied the idea you provide that "remove the loop for out_edges()" and "record edge iterator instead of visited nodes", but I probably not fully understand your idea:
1. I don't think we can totally get rid of the `work_stack`, because overall this is a stack-based algorithm and we need a stack to help us roll back, that's what `work_stack` does.
2. I prefer to to have `current_neighbor` as a `stack` rather than `vector`, then it will expand and shrink just as what `work_stack` does at the same time. This could avoid handling the `index` when rollback, it's more natural I think.

wenxizhu replied:
Anyway, I commit the change just now. To give us more confidence, I wrote 3 test cases:
1. A->B->C;                    // The most naive case, single input with single output.
2. A->B->C; A->B->D;   // Need the capability to roll back.
3. A->B->C; A->C;         // Handle the non-tree case.

And I confirmed these 3 patterns could successfully recognized in my implementation now. Please see my code for details.

penpornk reviewed (changes requested):
Thanks for your quick edits! I think `CheckForNodeFusion` is still not working as intended. I might have missed some of your comments, so please let me know if there's anything I haven't addressed as well. :)

wenxizhu committed:
'op name' -> 'op type'

wenxizhu committed:
Add a comment: storing the output slots of input nodes.

wenxizhu committed:
New algorithm applied to "CheckForNodeFusion()", to cover the pattern "A->B->C;A->C".

wenxizhu commented:
@penpornk Sorry for the late response! Please see my comments for detailed updates.

penpornk reviewed (changes requested):
Thank you for the changes and sorry for my delay as well! I think we are getting close. I proposed some changes to your `CheckForNodeFusion` function. Please take a look if it's okay.

wenxizhu committed:
Revert a change: accidently comment a few lines.

wenxizhu committed:
Some modifications to "CheckForNodeFusion()":
1. check for match before push;
2. use "nodes" instead of "work_stack";
3. get rid of "fi_ptr";

wenxizhu committed:
Apply clang-format-3.8

wenxizhu commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -2735,12 +2736,13 @@ MklLayoutRewritePass::CheckForNodeFusion(Node* a) const {
     //
 
     std::stack<Node *, std::vector<Node *>> work_stack;
```
Got your idea. I've adopted your modifications, please have a look

wenxizhu commented:
@penpornk Your modifications adopted, and also clang-format-3.8 applied. 

And about the test cases, I'm afraid there is no way to add such test cases, because:
1. As you know, "transpose-removal" is a very constrained pattern, those 3 test cases won't take effect on this pattern;
2. What we need is actually some "dummy" patterns, which have "multiple input" and "multiple output", and don't exist in our existing "mlk_layout_pass.cc " code. Any attempt to add these kind of "dummy" patterns through testing code will be destructive to existing "mkl_layout_pass.cc" code, at least a bunch of interfaces have to be added and implemented.

So can we leave these test cases to be added when some day a real-world pattern is met? That does make more sense I think.

penpornk reviewed (changes requested):
Thank you for the changes! Some line-wraps clang-format did look weird. Could you please help fix it? After this we should be good to go.

penpornk commented:
> So can we leave these test cases to be added when some day a real-world pattern is met? That does make more sense I think.
I agree. Thank you!

wenxizhu committed:
Format a wired comment made by clang-format-3.8

wenxizhu commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -514,8 +514,10 @@ class MklLayoutRewritePass : public GraphOptimizationPass {
+    // first,
+    // for example, graph "A->B->C-D" and finfo_ is {A->B->C to ABC, A->B->C->D
+    // to ABCD},
     // since the first gets applied first, the final graph will be ABC->D.
```
Fixed.

wenxizhu commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -514,8 +514,10 @@ class MklLayoutRewritePass : public GraphOptimizationPass {
-    // for example, graph "A->B->C-D" and finfo_ is {A->B->C to ABC, A->B->C->D to ABCD},
+    // The fusion patterns in "finfo_" that show up first will get applied
+    // first,
+    // for example, graph "A->B->C-D" and finfo_ is {A->B->C to ABC, A->B->C->D
```
Fixed.

wenxizhu commented:
@penpornk those 2 wired comment lines fixed now.

penpornk reviewed (dismissed):
Thanks so much again for the PR and for your patience!

wenxizhu commented:
@penpornk Thank you!

penpornk commented:
Please fix the `clang-format` error. Thank you!

wenxizhu commented:
@penpornk Hmm... already fixed I think. Did I miss any conversation?

penpornk commented:
It's the `Experimental clang-format Check` at the bottom of this PR. Can you see the `details` link? If not, I can copy the error message to here.

wenxizhu commented:
@penpornk OK, now I see the log, fixed now. However I'm not sure I use the same clang-format version with yours. Please let me know if there is still problems.

wenxizhu committed:
Break very long lines into 2.

wenxizhu commented:
@penpornk I tried fix it by hand. Let's see if it can pass format-check this time.

penpornk reviewed (approved):
Thanks a lot! Let's try again. :)

penpornk commented:
Note: Tagging `ready-to-pull` because the 5 failing tests also failed in the recent commits.

penpornk commented:
@wt-huang Could you please help pull this PR? Thank you!

wenxizhu commented:
@penpornk hi, these failed checks are in-relevant with my code, right?

penpornk commented:
@wenxizhu I think so. I think our tests are acting funny right now. It keeps asking me to re-apply the test tag.

wenxizhu commented:
@penpornk OK, thank you. And BTW, do you know when this PR will be merged?

penpornk commented:
@wenxizhu Unfortunately, I don't. It depends on @wt-huang now.

wenxizhu commented:
@penpornk OK, that's fine.

harshini-gadige commented:
> @penpornk OK, thank you. And BTW, do you know when this PR will be merged?

Helping to merge this as @wt-huang is OOO.

penpornk commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -815,6 +893,119 @@ class MklLayoutRewritePass : public GraphOptimizationPass {
+            e->dst_input() == kPermTensorIndex) {
+          // we find the "perm" node, now try to retrieve its value.
+          const TensorProto* proto = nullptr;
+          CHECK_EQ(GetNodeAttr(perm_node->def(), "value", &proto).ok(), true);
```
```suggestion
          DCHECK(GetNodeAttr(perm_node->def(), "value", &proto).ok());
```

wenxizhu replied:
Done.

penpornk commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -2391,6 +2640,143 @@ MklLayoutRewritePass::CheckForNodeRewrite(const Node* n) const {
+  // Create node.
+  Node* new_node;
+  TF_CHECK_OK(nb.Finalize(&**g, &new_node));
+  CHECK_NOTNULL(new_node);
```
```suggestion
  DCHECK(new_node);
```

wenxizhu replied:
Done.

penpornk commented on tensorflow/core/graph/mkl_layout_pass.cc:
```diff
@@ -2391,6 +2640,143 @@ MklLayoutRewritePass::CheckForNodeRewrite(const Node* n) const {
+  for (const Edge* e : transpose_to_nchw->out_edges()) {
+    if (!e->IsControlEdge()) {
+      const int kTransposeWithMklOpOutputSlot = 0;
+      CHECK_NOTNULL((*g)->AddEdge(new_node, kTransposeWithMklOpOutputSlot,
```
```suggestion
      DCHECK((*g)->AddEdge(new_node, kTransposeWithMklOpOutputSlot,
```

penpornk reviewed (changes requested):
I have some more minor changes from the internal check. We are trying to use ```DCHECK``` instead of ```CHECK``` and won't allow any new ```CHECK```s in the code anymore. Could you please help edit this? Thank you very much!

penpornk committed:
Update tensorflow/core/graph/mkl_layout_pass.cc

Use "DCHECK()" instead of "CHECK_EQ".

Co-Authored-By: wenxizhu <wenxi.zhu@intel.com>

penpornk committed:
Update tensorflow/core/graph/mkl_layout_pass.cc

Use "DCHECK()" instead of "CHECK_EQ".

Co-Authored-By: wenxizhu <wenxi.zhu@intel.com>

penpornk committed:
Update tensorflow/core/graph/mkl_layout_pass.cc

Use "DCHECK()" instead of "CHECK_EQ".

Co-Authored-By: wenxizhu <wenxi.zhu@intel.com>

wenxizhu commented:
@penpornk @harshini-gadige changes done!

penpornk reviewed (approved):
@wenxizhu Thank you very much and sorry for the inconvenience! :)

penpornk reviewed (changes requested):
@wenxizhu Unfortunately the new changes need `clang-format` change too. Could you please re-apply `clang-format` to the files?

For example, the error I'm seeing in the test (at the bottom of the PR page) is
```
--- a/tensorflow/core/graph/mkl_layout_pass.cc
+++ b/tensorflow/core/graph/mkl_layout_pass.cc
@@ -2700,8 +2700,8 @@ Status MklLayoutRewritePass::FuseTransposeMklOpTranspose(
   for (const Edge* e : transpose_to_nchw->out_edges()) {
     if (!e->IsControlEdge()) {
       const int kTransposeWithMklOpOutputSlot = 0;
-      DCHECK((*g)->AddEdge(new_node, kTransposeWithMklOpOutputSlot,
-                                  e->dst(), e->dst_input()));
+      DCHECK((*g)->AddEdge(new_node, kTransposeWithMklOpOutputSlot, e->dst(),
+                           e->dst_input()));
     }
   }
```

wenxizhu committed:
Fix format for clang-format check.

penpornk reviewed (approved):
Thank you!
